torch_xla/csrc/xla_backend_impl.h

```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include <iostream>
+#include <string>
+
+#include "tensorflow/compiler/xla/xla_client/computation_client.h"
+#include "torch/csrc/lazy/backend/backend_interface.h"
+#include "torch_xla/csrc/device.h"
+
+namespace torch_xla {
+class XLAData : public torch::lazy::BackendData {
+ public:
+  XLAData(const torch::lazy::Shape& shape,
+          const torch::lazy::BackendDevice& device,
+          xla::ComputationClient::DataPtr xla_data)
+      : torch::lazy::BackendData(device, shape), xla_data_(xla_data) {}
+
+  // TODO set Device and torch::lazy_shape correctly
+  XLAData(xla::ComputationClient::DataPtr xla_data)
+      : torch::lazy::BackendData(ParseDeviceString(xla_data->device()),
+                                 torch::lazy::Shape()),
+        xla_data_(xla_data) {}
+
+  Handle GetHandle() override { return xla_data_->GetOpaqueHandle(); }
+
+  void Assign(const torch::lazy::BackendData& data) override {
+    // Assign should only be called to update the handle, no need
+    // to update device and shape.
+    XLAData new_xla_data = static_cast<const XLAData&>(data).xla_data_;
+    xla_data_->Assign(*(new_xla_data.xla_data().get()));
+  }
+
+  bool HasValue() const override { return xla_data_->HasValue(); }
+
+  xla::ComputationClient::DataPtr xla_data() { return xla_data_; }
+
+ private:
+  // TODO: Do we really need a Share_Ptr here?
```

Comment:
Just curious -- which `shared_ptr` are we talking about here?

Reply:
So we generally pass `torch::lazy::BackendDataPtr` around as a `shared_ptr` but within the `XLAData` (which inherited from `BackendData` ) we have another `shared_ptr` of `xla::ComputationClient::DataPtr`. This two layer of shared_ptr is a bit awkward. I was thinking about making the `xla::ComputationClient::DataPtr` but I can not do that due to we need to pass it to runtime (we can't have both runtime and `BackendDataPtr` to hold reference to a `unique_ptr`). I think I will keep it as a `shared_ptr` for now.